File: backend/scripts/scraper.js
#!/usr/bin/env node

/**
 * YouTube Video Scraper (Node.js version)
 * Scrapes YouTube videos using yt-dlp and stores them in PostgreSQL database
 * 
 * Requirements:
 * - yt-dlp must be installed: pip install yt-dlp
 * - Node.js with required packages
 */

const { execSync, spawn } = require('child_process');
const { query, testConnection, closePool } = require('../config/database');
const fs = require('fs');
const path = require('path');

class YouTubeScraper {
  constructor() {
    this.searchTerm = "Automation Challenges";
    this.maxResults = 20;
    this.videos = [];
  }

  async checkYtDlp() {
    try {
      console.log('🔍 Checking if yt-dlp is installed...');
      execSync('yt-dlp --version', { stdio: 'pipe' });
      console.log('✅ yt-dlp is installed');
      return true;
    } catch (error) {
      console.log('❌ yt-dlp is not installed');
      console.log('💡 Install it with: pip install yt-dlp');
      return false;
    }
  }

  async createTableIfNotExists() {
    try {
      const createTableQuery = `
        CREATE TABLE IF NOT EXISTS scraped_videos (
          id SERIAL PRIMARY KEY,
          video_id VARCHAR(50) UNIQUE NOT NULL,
          title TEXT NOT NULL,
          description TEXT,
          view_count BIGINT,
          duration INTEGER,
          channel VARCHAR(255),
          thumbnail_url TEXT,
          video_url TEXT NOT NULL,
          published_at TIMESTAMP,
          scraped_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
          UNIQUE(video_id)
        );
      `;
      
      await query(createTableQuery);
      console.log('✅ scraped_videos table created/verified successfully');
      return true;
    } catch (error) {
      console.error('❌ Failed to create table:', error.message);
      return false;
    }
  }

  async searchYouTubeVideos() {
    try {
      console.log(`🔍 Searching YouTube for: "${this.searchTerm}" (max ${this.maxResults} results)`);
      
      // Use yt-dlp to search for videos
      const searchUrl = `ytsearch${this.maxResults}:${this.searchTerm}`;
      const command = `yt-dlp --flat-playlist --print "%(id)s|%(title)s|%(uploader)s|%(duration)s|%(view_count)s|%(thumbnail)s" "${searchUrl}"`;
      
      const output = execSync(command, { encoding: 'utf8', timeout: 60000 });
      const lines = output.trim().split('\n').filter(line => line.trim());
      
      console.log(`✅ Found ${lines.length} videos from search`);
      return lines;
      
    } catch (error) {
      console.error('❌ YouTube search failed:', error.message);
      return [];
    }
  }

  async getVideoDetails(videoId) {
    try {
      const videoUrl = `https://www.youtube.com/watch?v=${videoId}`;
      
      // Get detailed video info
      const command = `yt-dlp --print "%(description)s|%(upload_date)s" "${videoUrl}"`;
      const output = execSync(command, { encoding: 'utf8', timeout: 30000 });
      const parts = output.trim().split('|');
      
      return {
        description: parts[0] || '',
        upload_date: parts[1] || ''
      };
      
    } catch (error) {
      console.error(`❌ Failed to get details for video ${videoId}:`, error.message);
      return { description: '', upload_date: '' };
    }
  }

  async insertVideoToDatabase(videoData) {
    try {
      const insertQuery = `
        INSERT INTO scraped_videos (
          video_id, title, description, view_count, duration, 
          channel, thumbnail_url, video_url, published_at
        ) VALUES (
          $1, $2, $3, $4, $5, $6, $7, $8,
          CASE WHEN $9 = '' THEN NULL ELSE to_timestamp($9, 'YYYYMMDD') END
        )
        ON CONFLICT (video_id) 
        DO UPDATE SET
          title = EXCLUDED.title,
          description = EXCLUDED.description,
          view_count = EXCLUDED.view_count,
          duration = EXCLUDED.duration,
          channel = EXCLUDED.channel,
          thumbnail_url = EXCLUDED.thumbnail_url,
          scraped_at = CURRENT_TIMESTAMP
        RETURNING id, video_id, title;
      `;
      
      const values = [
        videoData.video_id,
        videoData.title,
        videoData.description,
        videoData.view_count,
        videoData.duration,
        videoData.channel,
        videoData.thumbnail_url,
        videoData.video_url,
        videoData.upload_date
      ];
      
      const result = await query(insertQuery, values);
      
      if (result.rows && result.rows.length > 0) {
        console.log(`✅ Inserted/Updated video: ${result.rows[0].title}`);
        return true;
      }
      
      return false;
      
    } catch (error) {
      console.error(`❌ Database error for video ${videoData.video_id}:`, error.message);
      return false;
    }
  }

  async runScraping() {
    try {
      console.log('🚀 Starting YouTube video scraping process');
      
      // Test database connection
      await testConnection();
      console.log('✅ Database connection successful');
      
      // Check yt-dlp
      if (!(await this.checkYtDlp())) {
        return false;
      }
      
      // Create table if needed
      if (!(await this.createTableIfNotExists())) {
        return false;
      }
      
      // Search for videos
      const searchResults = await this.searchYouTubeVideos();
      if (searchResults.length === 0) {
        console.log('⚠️ No videos found to scrape');
        return false;
      }
      
      // Process each video
      let successfulInserts = 0;
      let failedInserts = 0;
      
      for (let i = 0; i < searchResults.length; i++) {
        const line = searchResults[i];
        const parts = line.split('|');
        
        if (parts.length < 6) {
          console.log(`⚠️ Skipping malformed video data: ${line}`);
          failedInserts++;
          continue;
        }
        
        const [videoId, title, channel, duration, viewCount, thumbnail] = parts;
        
        console.log(`📹 Processing video ${i + 1}/${searchResults.length}: ${title}`);
        
        // Get additional video details
        const details = await this.getVideoDetails(videoId);
        
        const videoData = {
          video_id: videoId,
          title: title,
          description: details.description,
          view_count: parseInt(viewCount) || 0,
          duration: parseInt(duration) || 0,
          channel: channel,
          thumbnail_url: thumbnail,
          video_url: `https://www.youtube.com/watch?v=${videoId}`,
          upload_date: details.upload_date
        };
        
        // Insert into database
        if (await this.insertVideoToDatabase(videoData)) {
          successfulInserts++;
        } else {
          failedInserts++;
        }
        
        // Small delay to avoid overwhelming the system
        await new Promise(resolve => setTimeout(resolve, 1000));
      }
      
      // Summary
      console.log('🎉 Scraping completed!');
      console.log(`📊 Results: ${successfulInserts} successful, ${failedInserts} failed`);
      
      return true;
      
    } catch (error) {
      console.error('❌ Scraping process failed:', error.message);
      return false;
    }
  }
}

// Main execution
async function main() {
  const scraper = new YouTubeScraper();
  
  try {
    const success = await scraper.runScraping();
    
    if (success) {
      console.log('✅ Scraping process completed successfully');
      process.exit(0);
    } else {
      console.log('❌ Scraping process failed');
      process.exit(1);
    }
    
  } catch (error) {
    console.error('❌ Unexpected error:', error.message);
    process.exit(1);
  } finally {
    await closePool();
  }
}

// Run the script
if (require.main === module) {
  main().catch(console.error);
}

module.exports = YouTubeScraper;

File: scripts/scraper.py
#!/usr/bin/env python3
"""
YouTube Video Scraper
Scrapes YouTube videos using yt-dlp and stores them in PostgreSQL database
"""

import os
import sys
import yt_dlp
import psycopg2
from psycopg2.extras import RealDictCursor
from psycopg2 import sql
from dotenv import load_dotenv
from urllib.parse import urlparse
import logging

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class YouTubeScraper:
    def __init__(self):
        """Initialize the scraper with database connection"""
        load_dotenv()
        self.db_connection = None
        self.search_term = "Automation Challenges"
        self.max_results = 20
        
    def connect_database(self):
        """Connect to PostgreSQL database using credentials from .env"""
        try:
            # Parse DATABASE_URL if available, otherwise use individual variables
            database_url = os.getenv('DATABASE_URL')
            
            if database_url:
                logger.info("Using DATABASE_URL for connection")
                self.db_connection = psycopg2.connect(database_url, cursor_factory=RealDictCursor)
            else:
                # Fallback to individual environment variables
                logger.info("Using individual database credentials")
                self.db_connection = psycopg2.connect(
                    host=os.getenv('DB_HOST', 'localhost'),
                    port=os.getenv('DB_PORT', '5432'),
                    database=os.getenv('DB_NAME', 'workflow_automation'),
                    user=os.getenv('DB_USER', 'postgres'),
                    password=os.getenv('DB_PASSWORD', 'password'),
                    cursor_factory=RealDictCursor
                )
            
            logger.info("✅ Successfully connected to PostgreSQL database")
            return True
            
        except Exception as e:
            logger.error(f"❌ Failed to connect to database: {e}")
            return False
    
    def create_table_if_not_exists(self):
        """Create scraped_videos table if it doesn't exist"""
        try:
            cursor = self.db_connection.cursor()
            
            create_table_query = """
            CREATE TABLE IF NOT EXISTS scraped_videos (
                id SERIAL PRIMARY KEY,
                video_id VARCHAR(50) UNIQUE NOT NULL,
                title TEXT NOT NULL,
                description TEXT,
                view_count BIGINT,
                duration INTEGER,
                channel VARCHAR(255),
                thumbnail_url TEXT,
                video_url TEXT NOT NULL,
                published_at TIMESTAMP,
                scraped_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                UNIQUE(video_id)
            );
            """
            
            cursor.execute(create_table_query)
            self.db_connection.commit()
            cursor.close()
            
            logger.info("✅ scraped_videos table created/verified successfully")
            return True
            
        except Exception as e:
            logger.error(f"❌ Failed to create table: {e}")
            return False
    
    def search_youtube_videos(self):
        """Search YouTube for videos using yt-dlp"""
        try:
            # Configure yt-dlp options for search
            ydl_opts = {
                'quiet': True,
                'no_warnings': True,
                'extract_flat': True,  # Don't download, just extract metadata
                'max_downloads': self.max_results,
                'ignoreerrors': True,  # Continue on errors
            }
            
            # Search URL for the term
            search_url = f"ytsearch{self.max_results}:{self.search_term}"
            
            logger.info(f"🔍 Searching YouTube for: '{self.search_term}' (max {self.max_results} results)")
            
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                # Extract search results
                search_results = ydl.extract_info(search_url, download=False)
                
                if not search_results or 'entries' not in search_results:
                    logger.warning("No search results found")
                    return []
                
                videos = []
                for entry in search_results['entries']:
                    if entry:  # Skip None entries
                        videos.append(entry)
                
                logger.info(f"✅ Found {len(videos)} videos from search")
                return videos
                
        except Exception as e:
            logger.error(f"❌ YouTube search failed: {e}")
            return []
    
    def extract_video_details(self, video_entry):
        """Extract detailed video information using yt-dlp"""
        try:
            video_id = video_entry.get('id', '')
            video_url = f"https://www.youtube.com/watch?v={video_id}"
            
            # Get detailed video info
            ydl_opts = {
                'quiet': True,
                'no_warnings': True,
                'ignoreerrors': True,
            }
            
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                detailed_info = ydl.extract_info(video_url, download=False)
                
                if not detailed_info:
                    return None
                
                # Extract thumbnail URL (get highest quality)
                thumbnails = detailed_info.get('thumbnails', [])
                thumbnail_url = ''
                if thumbnails:
                    # Sort by resolution and get the highest quality
                    thumbnails.sort(key=lambda x: x.get('height', 0) * x.get('width', 0), reverse=True)
                    thumbnail_url = thumbnails[0].get('url', '')
                
                # Extract data
                video_data = {
                    'video_id': video_id,
                    'title': detailed_info.get('title', ''),
                    'description': detailed_info.get('description', ''),
                    'view_count': detailed_info.get('view_count', 0),
                    'duration': detailed_info.get('duration', 0),
                    'channel': detailed_info.get('uploader', ''),
                    'thumbnail_url': thumbnail_url,
                    'video_url': video_url,
                    'published_at': detailed_info.get('upload_date', ''),
                }
                
                return video_data
                
        except Exception as e:
            logger.error(f"❌ Failed to extract details for video {video_entry.get('id', 'unknown')}: {e}")
            return None
    
    def insert_video_to_database(self, video_data):
        """Insert video data into PostgreSQL database"""
        try:
            cursor = self.db_connection.cursor()
            
            # Prepare the insert query with ON CONFLICT handling
            insert_query = """
            INSERT INTO scraped_videos (
                video_id, title, description, view_count, duration, 
                channel, thumbnail_url, video_url, published_at
            ) VALUES (
                %(video_id)s, %(title)s, %(description)s, %(view_count)s, %(duration)s,
                %(channel)s, %(thumbnail_url)s, %(video_url)s, 
                CASE WHEN %(published_at)s = '' THEN NULL ELSE to_timestamp(%(published_at)s, 'YYYYMMDD') END
            )
            ON CONFLICT (video_id) 
            DO UPDATE SET
                title = EXCLUDED.title,
                description = EXCLUDED.description,
                view_count = EXCLUDED.view_count,
                duration = EXCLUDED.duration,
                channel = EXCLUDED.channel,
                thumbnail_url = EXCLUDED.thumbnail_url,
                scraped_at = CURRENT_TIMESTAMP
            RETURNING id, video_id, title;
            """
            
            cursor.execute(insert_query, video_data)
            result = cursor.fetchone()
            self.db_connection.commit()
            cursor.close()
            
            if result:
                logger.info(f"✅ {'Inserted' if 'INSERT' in str(cursor.statusmessage) else 'Updated'} video: {result['title']}")
                return True
            else:
                logger.warning(f"⚠️ No result returned for video: {video_data['title']}")
                return False
                
        except psycopg2.IntegrityError as e:
            logger.warning(f"⚠️ Integrity error for video {video_data['video_id']}: {e}")
            return False
        except Exception as e:
            logger.error(f"❌ Database error for video {video_data['video_id']}: {e}")
            return False
    
    def run_scraping(self):
        """Main function to execute the scraping process"""
        try:
            logger.info("🚀 Starting YouTube video scraping process")
            
            # Connect to database
            if not self.connect_database():
                return False
            
            # Create table if needed
            if not self.create_table_if_not_exists():
                return False
            
            # Search for videos
            search_results = self.search_youtube_videos()
            if not search_results:
                logger.warning("No videos found to scrape")
                return False
            
            # Process each video
            successful_inserts = 0
            failed_inserts = 0
            
            for i, video_entry in enumerate(search_results, 1):
                logger.info(f"📹 Processing video {i}/{len(search_results)}: {video_entry.get('title', 'Unknown')}")
                
                # Extract detailed video information
                video_data = self.extract_video_details(video_entry)
                
                if video_data:
                    # Insert into database
                    if self.insert_video_to_database(video_data):
                        successful_inserts += 1
                    else:
                        failed_inserts += 1
                else:
                    failed_inserts += 1
                    logger.warning(f"⚠️ Skipped video due to extraction failure")
            
            # Summary
            logger.info(f"🎉 Scraping completed!")
            logger.info(f"📊 Results: {successful_inserts} successful, {failed_inserts} failed")
            
            return True
            
        except Exception as e:
            logger.error(f"❌ Scraping process failed: {e}")
            return False
        
        finally:
            # Close database connection
            if self.db_connection:
                self.db_connection.close()
                logger.info("🔌 Database connection closed")

def main():
    """Main entry point"""
    try:
        scraper = YouTubeScraper()
        success = scraper.run_scraping()
        
        if success:
            logger.info("✅ Scraping process completed successfully")
            sys.exit(0)
        else:
            logger.error("❌ Scraping process failed")
            sys.exit(1)
            
    except KeyboardInterrupt:
        logger.info("🛑 Scraping interrupted by user")
        sys.exit(1)
    except Exception as e:
        logger.error(f"❌ Unexpected error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
